hudi-client/src/main/java/org/apache/hudi/index/bloom/HoodieSimpleIndex.java

```diff
@@ -0,0 +1,211 @@
+    JavaRDD<Tuple2<String, String>> fileInfoList =
+        jsc.parallelize(loadAllFilesForPartitions(affectedPartitionPathList, jsc, hoodieTable));
+
+    return findMatchingFilesForRecordKeys(fileInfoList, partitionRecordKeyPairRDD, hoodieTable);
+  }
+
+  /**
+   * Load all involved files as <Partition, filename> pair RDD.
+   */
+  @VisibleForTesting
+  List<Tuple2<String, String>> loadAllFilesForPartitions(List<String> partitions, final JavaSparkContext jsc,
+                                                         final HoodieTable hoodieTable) {
+
+    // Obtain the latest data files from all the partitions.
+    List<Pair<String, String>> partitionPathFileIDList = loadLatestDataFilesForAllPartitions(partitions,
+        jsc, hoodieTable);
+    return partitionPathFileIDList.stream()
+        .map(pf -> new Tuple2<>(pf.getKey(), pf.getValue())).collect(toList());
+  }
+
+  /**
+   * Find <HoodieKey, HoodieRecordLocation> for all incoming HoodieKeys
+   */
+  private JavaPairRDD<HoodieKey, HoodieRecordLocation> findMatchingFilesForRecordKeys(JavaRDD<Tuple2<String, String>> partitionToFileIndexInfo,
+                                                                                      JavaPairRDD<String, String> partitionRecordKeyPairRDD, HoodieTable hoodieTable) {
+    // Step 1: for each <partition, fileId> pair, co locate records to be searched for
+    JavaPairRDD<String, String> partitionFileIndexInfoPairRDD = partitionToFileIndexInfo.mapToPair(entry -> new Tuple2<>(entry._1, entry._2));
+    JavaPairRDD<String, List<String>> partitionToRecords = partitionRecordKeyPairRDD.groupByKey().mapToPair(entry -> new Tuple2(entry._1, Lists.newArrayList(entry._2)));
+    JavaPairRDD<String, Tuple2<String, List<String>>> partitionToFileIdAndRecordsPairRDD = partitionFileIndexInfoPairRDD.join(partitionToRecords);
+
+    // Step 2: For each partition, fileId, List<recordKeys> Triplet -> Fetch RDD of triplets ( Tuple2< HoodieKey, HoodieRecordLocation >)
+
+    /*Option1:
+    return partitionToFileIdAndRecordsPairRDD.flatMapToPair(
+        (PairFlatMapFunction<Tuple2<String, Tuple2<String, List<String>>>, HoodieKey, HoodieRecordLocation>) stringTuple2Tuple2 -> {
+          HoodieDataFile latestDataFile = getLatestDataFile(hoodieTable, Pair.of(stringTuple2Tuple2._1, stringTuple2Tuple2._2._1));
+          List<Pair<HoodieKey, HoodieRecordLocation>> resultList = ParquetUtils.fetchRecordKeyPartitionPathFromParquet(hoodieTable.getHadoopConf(), new Path(latestDataFile.getPath()),
+              latestDataFile.getCommitTime(), stringTuple2Tuple2._2._1, stringTuple2Tuple2._2._2);
+          List<Tuple2<HoodieKey, HoodieRecordLocation>> toReturn = new ArrayList<>();
+          resultList.forEach(rec -> toReturn.add(new Tuple2<>(rec.getLeft(), rec.getRight())));
+          return toReturn.iterator();
+        });*/
+
+    // Option2:
+    return partitionToFileIdAndRecordsPairRDD.mapPartitions(
```

Comment:
Since the incoming records are files, I am not sure how much value we get in introducing a LazyIterator. I have both options listed here(using mapPartitions, vs lazy iterator). Let me know your thoughts. Interested in understanding the nuances.